### PR TITLE
New-app: Expose ports specified in source Dockerfile

### PR DESCRIPTION
**Pull request by csrwng**
Fixes https://github.com/openshift/origin/issues/3192

### CODE REVIEW

**Comment by csrwng** [comment]
@deads2k fyi -- adding a test

**Comment by csrwng** [comment]
test added

**Comment by csrwng** [comment]
[test]

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -418,6 +418,19 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
 						return nil, err
 					}
 				}
+				// Append any exposed ports from Dockerfile to input image
```
Comment (deads2k) [inline pkg/generate/app/cmd/newapp.go]: An observation about the code.  It's really difficult to know which parts of the objects are supposed to be valid at any given point in time.  The `ComponentReference` holds a `ComponentInput` which could contain this information as early as the `componentsForRepos` step, but the `ComponentInput` gets stomped again during the `Resolve` step.  

Having this change here means that even after `Resolve` is finished, the `ComponentInput` is incomplete.  I *think* the `ComponentInput` is getting mutated here via  series of navigated references, but I'm not completely certain.  If so, it feels a little weird to have something called `buildPipelines` change an input object.

Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: > Having this change here means that even after Resolve is finished, the ComponentInput is incomplete. I think the ComponentInput is getting mutated here via series of navigated references, but I'm not completely certain. If so, it feels a little weird to have something called buildPipelines change an input object.

Not quite. So the input image ref is created here in this method (using ref.Input().Match as the input) https://github.com/csrwng/origin/blob/newapp_dockerfile_ports/pkg/generate/app/cmd/newapp.go#L403, and we're simply mutating that created imageref.

Having said that, I agree that it is hard to follow what objects are available at what time. A restructuring and maybe a refactor to a more functional style of code here with immutable inputs/outputs would be great. But that won't happen right now :-)

Comment (kargakis) [inline pkg/generate/app/cmd/newapp.go]: Port detection was previously happening while in source detection (removed by [this PR](https://github.com/openshift/origin/commit/5ab34d377db4710589d0b3f15cebea23e1f6500f). We should restore it somewhere there and not so late in the new-app flow.

Comment (deads2k) [inline pkg/generate/app/cmd/newapp.go]: > Port detection was previously happening while in source detection (removed by this PR). We should restore it somewhere there and not so late in the new-app flow.

An issue describing what we'd like to achieve in a refactor would be sufficient for me.  I tried to make this modification myself, but after spending an hour in the code, I couldn't figure out where to do it.  My changes kept getting stomped.

Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: > Port detection was previously happening while in source detection (removed by this PR). We should restore it somewhere there and not so late in the new-app flow.

We still get the data during source detection (populated Dockerfile). However, the flow is different now since we do a resolve step which wasn't happening before.

Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: > An issue describing what we'd like to achieve in a refactor would be sufficient for me. I tried to make this modification myself, but after spending an hour in the code, I couldn't figure out where to do it. My changes kept getting stomped.

https://github.com/openshift/origin/issues/3215

**Comment by deads2k** [comment]
This isn't handling multiple ports exposed on the same line: 
```
EXPOSE 111/udp 2049/tcp
```

is producing
```
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: null
    name: docker-nfs-server
  spec:
    portalIP: ""
    ports:
    - name: docker-nfs-server-udp-111
      nodePort: 0
      port: 111
      protocol: UDP
      targetPort: 111
    selector:
      deploymentconfig: docker-nfs-server
  status:
    loadBalancer: {}

```

**Comment by csrwng** [comment]
It's a feature :)
We only expose a single port as a service, by design.

**Comment by csrwng** [comment]
The container in the deployment config, however, should include all ports

**Comment by deads2k** [comment]
> We only expose a single port as a service, by design.

Why would I want that?  If I have explicitly said in my `Dockerfile` that I want to expose multiple ports, doesn't it seem reasonable to expose all of them by default?

**Comment by deads2k** [comment]
Take my case as a for instance.  nfs requires more than one port, so I've exposed them in my Dockerfile, but I can't use new-app because it only exposes one.

**Comment by csrwng** [comment]
The thinking is that you could expose more if you like via the ```oc expose``` command - per @smarterclayton

**Comment by deads2k** [comment]
> The thinking is that you could expose more if you like via the oc expose command - per

That's certainly a recovery option, but I think the default should be full exposure.  The author exposed them for a reason and it's easier for people to edit and delete than to edit and create.

**Comment by csrwng** [comment]
rebased. @smarterclayton @deads2k - this one should be ready to go in

**Comment by deads2k** [comment]
lgtm.  Will take multiple ports as a separate pull.  [merge]

**Comment by csrwng** [comment]
trying again... [merge]

**Comment by deads2k** [comment]
re[merge]

**Comment by csrwng** [comment]
hmm, haven't seen this one before... 
```
/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/atoms/error.js:113
  var template = new Error(this.message);
                 ^
UnknownError: Error communicating with the remote browser. It may have died.
Build info: version: '2.43.1', revision: '5163bce', time: '2014-09-10 16:27:33'
Driver info: driver.version: PhantomJSDriver
```

**Comment by liggitt** [comment]
Web driver flake. There's an issue open to investigate replacing with headless chrome

**Comment by csrwng** [comment]
[merge] again
